### PR TITLE
Add rebatch method for Dataset

### PR DESCRIPTION
**Pull request by yongtang**
This PR adds rebatch method for Dataset where
```
dataset.apply(rebatch(n)) = dataset.unbatch().batch(n)
```

The motivation for rebatch is that there are situations we read the data in
big batches but then we want to adjust the batch size to fit differnet
scenarios.

This is part of #382.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow_io/core/kernels/rebatch_dataset_op.cc [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        : DatasetBase(DatasetContext(ctx)),
+          batch_size_(batch_size),
+          batch_mode_(batch_mode),
+          input_(input) {
+      input_->Ref();
+
+      const auto& input_shapes = input_->output_shapes();
+      output_shapes_.reserve(input_shapes.size());
+      // Always set the first dim as None unless batch_mode is specified.
+      for (const auto& input_shape : input_shapes) {
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/kernels/rebatch_dataset_op.cc]: @feihugis Done. PR updated.

File: tensorflow_io/core/kernels/rebatch_dataset_op.cc [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+                out_tensors->emplace_back(ctx->allocator({}), tensors_[i].dtype(), shape);
+              }
+            }
+            if (out_tensors->size() != tensors_.size()) {
+              return errors::InvalidArgument("number tensors should match previous one, ", tensors_.size(), " vs. ", out_tensors->size());
+            }
+            int64 chunk_to_read = (current_batch_size_ - current_index_) < (dataset()->batch_size_ - chunk_read) ? (current_batch_size_ - current_index_) : (dataset()->batch_size_ - chunk_read);
+            for (int i = 0; i < tensors_.size(); ++i) {
+              // TODO: concurrent copy?
+              for (int64 r = 0; r < chunk_to_read; r++) {
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/kernels/rebatch_dataset_op.cc]: Done.

File: tests/test_text_eager.py [hunk 3]
```diff
@@ -53,6 +54,13 @@ def test_text_input():
       i += 1
   assert i == len(lines)
 
+  rebatch_dataset = text_dataset.apply(core_io.rebatch(5))
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tests/test_text_eager.py]: Done. Additional tests added.

File: tensorflow_io/core/ops/core_ops.cc [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/framework/common_shape_fns.h"
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/shape_inference.h"
+
+namespace tensorflow {
+
+REGISTER_OP("RebatchDataset")
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/ops/core_ops.cc]: The `batch_mode` input could take a string to specify the batch mode:
- keep: leave the reminder as is.
- drop: drop the reminder
- pad: pad the reminder.

File: tensorflow_io/core/kernels/rebatch_dataset_op.cc [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                    tensors_[0].dim_size(0), " but component ", i,
+                    " had size, ", tensors_[i].dim_size(0), ".");
+              }
+            }
+            current_batch_size_ = tensors_[0].dim_size(0);
+          }
+        }
+        // Finally, resize if needed
+        if (chunk_read > 0) {
+          if (chunk_read < dataset()->batch_size_) {
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/kernels/rebatch_dataset_op.cc]: Updated. `keep`, `drop`, and `pad` modes have been added.

File: tensorflow_io/core/kernels/rebatch_dataset_op.cc [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+              for (size_t i = 0; i < tensors_.size(); ++i) {
+                TensorShape shape = tensors_[i].shape();
+                shape.RemoveDim(0);
+                elements.emplace_back(ctx->allocator({}), tensors_[i].dtype(), shape);
+                shape.InsertDim(0, dataset()->batch_size_);
+                out_tensors->emplace_back(ctx->allocator({}), tensors_[i].dtype(), shape);
+              }
+            }
+            if (out_tensors->size() != tensors_.size()) {
+              return errors::InvalidArgument("number tensors should match previous one, ", tensors_.size(), " vs. ", out_tensors->size());
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/kernels/rebatch_dataset_op.cc]: In TensorFlow, at one point the C++ style was enforced with `clang-format`. The issue with `clang-format` was that different versions of `clang-format` have different styles so it is really not easy to figure out which one is the right one. TensorFlow dropped the C++ style check later.

I think we could leave the C++ style check alone until we find a clang-format version that stabilize.

**Comment by yongtang** [comment]
@feihugis  Thanks for the review. The `batch_mode` takes `keep`, `drop`, `pad` modes to decide what to do when reminder surface.

Also the name of the C++ class has been changed to `AdjustBatchDataset`.

For python function name I really think `"rebatch"` makes plenty of sense. I will just leave as is. In the future if this Dataset ops is to be added to TensorFlow core repo then we could rethink the name I think.

File: tensorflow_io/core/kernels/rebatch_dataset_op.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow_io/core/kernels/rebatch_dataset_op.cc]: Got it. Thanks!

File: tensorflow_io/core/kernels/rebatch_dataset_op.cc [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                   DatasetBase** output) override {
+    int64 batch_size = 0;
+    OP_REQUIRES_OK(ctx,
+                   ParseScalarArgument<int64>(ctx, "batch_size", &batch_size));
+    OP_REQUIRES(
+        ctx, batch_size > 0,
+        errors::InvalidArgument("Batch size must be greater than zero."));
+
+    string batch_mode = "";
+    OP_REQUIRES_OK(ctx,
```
Comment (feihugis) [inline tensorflow_io/core/kernels/rebatch_dataset_op.cc]: minor: do we need to check if the input batch_mode is valid?

File: tensorflow_io/core/kernels/rebatch_dataset_op.cc [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+                (*out_tensors)[i] = std::move(value_tensor);
+              }
+            // "drop" the reminder
+            } else if (dataset()->batch_mode_ == "drop") {
+              out_tensors->clear();
+              input_impl_.reset();
+              *end_of_sequence = true;
+              return Status::OK();
+            }
+            // otherwise "pad" means keep the size
```
Comment (feihugis) [inline tensorflow_io/core/kernels/rebatch_dataset_op.cc]: Just remind that `pad` is not implemented yet.

**Pull request merged**